src/Microsoft.Framework.DesignTimeHost/PluginHandler.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.Framework.DesignTimeHost.Models.IncomingMessages;
+using Microsoft.Framework.Runtime;
+using Microsoft.Framework.Runtime.Common.DependencyInjection;
+
+namespace Microsoft.Framework.DesignTimeHost
+{
+    public class PluginHandler
+    {
+        private readonly Action<object> _sendMessageMethod;
+        private readonly IServiceProvider _hostServices;
+        private readonly IDictionary<string, IPlugin> _plugins;
+
+        public PluginHandler(IServiceProvider hostServices, Action<object> sendMessageMethod)
+        {
+            _sendMessageMethod = sendMessageMethod;
+            _hostServices = hostServices;
+            _plugins = new Dictionary<string, IPlugin>(StringComparer.Ordinal);
+        }
+
+        public void ProcessMessage(PluginMessage data, IAssemblyLoadContext assemblyLoadContext)
+        {
+            switch (data.MessageName)
+            {
+                case "RegisterPlugin":
+                    ProcessRegisterMessage(data, assemblyLoadContext);
+                    break;
+                case "UnregisterPlugin":
+                    ProcessUnregisterMessage(data);
+                    break;
+                case "PluginMessage":
+                    ProcessPluginMessage(data);
+                    break;
+            }
+        }
+
+        private void ProcessPluginMessage(PluginMessage data)
+        {
+            IPlugin plugin;
+            if (_plugins.TryGetValue(data.PluginId, out plugin))
+            {
+                plugin.ProcessMessage(data.Data);
+            }
+            else
+            {
+                throw new InvalidOperationException(
+                    Resources.FormatPlugin_UnregisteredPluginIdCannotReceiveMessages(data.PluginId));
+            }
+        }
+
+        private void ProcessUnregisterMessage(PluginMessage data)
+        {
+            if (_plugins.ContainsKey(data.PluginId))
+            {
+                _plugins.Remove(data.PluginId);
+            }
+            else
+            {
+                throw new InvalidOperationException(
+                    Resources.FormatPlugin_UnregisteredPluginIdCannotUnregister(data.PluginId));
+            }
+        }
+
+        private void ProcessRegisterMessage(PluginMessage data, IAssemblyLoadContext assemblyLoadContext)
+        {
+            var pluginId = data.PluginId;
+            var registerData = data.Data.ToObject<PluginRegisterData>();
+            var assembly = assemblyLoadContext.Load(registerData.AssemblyName);
+            var pluginType = assembly.GetType(registerData.TypeName);
+
+            if (pluginType != null)
```

Comment:
That's thrown by the `.Load`, had a review comment about that in the initial iterations but @davidfowl seemed to think it was ok as is.